metagraph/src/graph/representation/succinct/dbg_succinct_range.cpp

```diff
@@ -0,0 +1,864 @@
+    }
+}
+
+size_t DBGSuccinctRange::outdegree(node_index node) const {
+    if (node < offset_)
+        return dbg_succ_.outdegree(node);
+
+    auto [edge_range, is_sink] = fetch_edge_range(node);
+    if (is_sink)
+        return 1;
+
+    // TODO: better implementation: count the number of symbols in W in the range
+    //       reimplement with sdsl::wt_pc::interval_symbols
+    size_t count = 0;
+    call_outgoing_kmers(node, [&count](auto, auto) { ++count; });
+    return count;
+}
+
+bool DBGSuccinctRange::has_single_outgoing(node_index node) const {
+    if (node < offset_)
+        return dbg_succ_.has_single_outgoing(node);
+
+    auto [edge_range, is_sink] = fetch_edge_range(node);
+    auto [first, last, offset] = edge_range;
+    const auto &boss_graph = dbg_succ_.get_boss();
+
+    if (is_sink)
+        return true;
+
+    assert(boss_graph.get_last(first));
+
+    // check if there is only one node in the range, with only one edge
+    return first == last && boss_graph.get_last(first - 1);
+}
+
+bool DBGSuccinctRange::has_multiple_outgoing(node_index node) const {
+    if (node < offset_)
+        return dbg_succ_.has_multiple_outgoing(node);
+
+    auto [edge_range, is_sink] = fetch_edge_range(node);
+    if (is_sink)
+        return false;
+
+    auto [first, last, offset] = edge_range;
+    assert(dbg_succ_.get_boss().get_last(first));
+
+    // check if there are multiple nodes in the range, or if the node has
+    // multiple outgoing edges
+    return first != last || !dbg_succ_.get_boss().get_last(first - 1);
+}
+
+size_t DBGSuccinctRange::indegree(node_index node) const {
+    if (node < offset_)
+        return dbg_succ_.indegree(node);
+
+    auto [edge_range, is_sink] = fetch_edge_range(node);
+    if (!is_sink)
+        return 1;
+
+    // TODO: more efficient
+    size_t count = 0;
+    call_incoming_kmers(node, [&count](auto, auto) { ++count; });
+    return count;
+}
+
+bool DBGSuccinctRange::has_no_incoming(node_index node) const {
+    return node < offset_ && dbg_succ_.has_no_incoming(node);
+}
+
+bool DBGSuccinctRange::has_single_incoming(node_index node) const {
+    if (node < offset_)
+        return dbg_succ_.has_single_incoming(node);
+
+    auto [edge_range, is_sink] = fetch_edge_range(node);
+
+    if (!is_sink)
+        return true;
+
+    // TODO: more efficient
+    return indegree(node) == 1;
+}
+
+DBGSuccinctRange::node_index DBGSuccinctRange
+::kmer_to_node(const boss::BOSS::TAlphabet *begin,
+               const boss::BOSS::TAlphabet *end,
+               bool require_exact_length) const {
+    const auto &boss_graph = dbg_succ_.get_boss();
+    assert(begin <= end);
+    assert(boss_graph.get_k() >= static_cast<size_t>(end - begin));
+
+    if (std::find(begin, end, boss_graph.alph_size) != end)
+        return 0;
+
+    auto [first, last, seq_it] = boss_graph.index_range(begin, end);
+
+    const boss::BOSS::TAlphabet *min_match_pos = require_exact_length ? end : begin + 1;
```

Comment:
for some reason, setting `begin + min_match_length_` instead of `begin + 1` almost doubles the run time